lua/ttt2/libraries/vguihandler.lua

```diff
@@ -0,0 +1,236 @@
+	vguihandler.callback.frame = {}
+end
+
+---
+-- Returns if a menu is currently hidden
+-- @return boolean The hide state
+-- @realm client
+function vguihandler.IsHidden()
+	return IsValid(vguihandler.menuCache.hidden) or false
+end
+
+---
+-- Call this function to unhide a frame, does nothing if no
+-- frame was hidden. Also calls a callback function if set.
+-- @realm client
+function vguihandler.UnhideFrame()
+	if not IsValid(vguihandler.menuCache.hidden) then return end
+
+	if isfunction(vguihandler.callback.hidden["unhide"]) then
+		local shouldCancel = vguihandler.callback.hidden["unhide"](vguihandler.menuCache.hidden) == false
+
+		if shouldCancel then return end
+	end
+
+	-- close frame if there was one opened while being hidden
+	vguihandler.CloseFrame()
+
+	-- if closing failed (can be stopped by callback), hiding
+	-- also fails
+	if IsValid(vguihandler.menuCache.frame) then return end
+
+	vguihandler.menuCache.hidden:SetDeleteOnClose(true)
+	vguihandler.menuCache.hidden:SetVisible(true)
+
+	-- move frame from hidden to active frame
+	vguihandler.menuCache.frame = vguihandler.menuCache.hidden
+	vguihandler.menuCache.hidden = nil
+
+	vguihandler.callback.frame = vguihandler.callback.hidden
+	vguihandler.callback.hidden = {}
+end
+
+---
+-- Returns if a menu is open or not
+-- @return boolean True if a menu is open
+-- @realm client
+function vguihandler.IsOpen()
+	return IsValid(vguihandler.menuCache.frame) or false
+end
+
+local function InternalUpdateVSkinSetting(name, panel)
+	if name == "blur" then
+		panel:SetBackgroundBlur(vskin.ShouldBlurBackground())
+
+		panel:InvalidateLayout()
+	elseif name == "skin" then
+		panel:InvalidateLayout()
+	elseif name == "language" then
+		panel:InvalidateLayout()
+	elseif name == "general_rebuild" then
+		panel:InvalidateLayout()
+	end
+end
+
+---
+-- Should be called when a specific vskin variable is changed
+-- so that the complete vgui element is redone
+-- @param string name The name of the changed setting
+-- @internal
+-- @realm client
+function vguihandler.UpdateVSkinSetting(name)
+	if IsValid(vguihandler.menuCache.frame) then
+		InternalUpdateVSkinSetting(name, vguihandler.menuCache.frame)
+	end
+
+	if IsValid(vguihandler.menuCache.hidden) then
+		InternalUpdateVSkinSetting(name, vguihandler.menuCache.hidden)
+	end
+end
+
+---
+-- Rebuilds the whole menu without a specific changed settings
+-- @realm client
+function vguihandler.Rebuild()
+	if isfunction(vguihandler.callback.frame["rebuild"]) then
+		vguihandler.callback.frame["rebuild"](vguihandler.menuCache.frame)
+	end
+
+	if isfunction(vguihandler.callback.hidden["rebuild"]) then
+		vguihandler.callback.hidden["rebuild"](vguihandler.menuCache.hidden)
+	end
+
+	vguihandler.UpdateVSkinSetting("general_rebuild")
+end
+
+---
+-- Registers a callback function for a specific event
+-- @param string name The name of the event
+-- @param function fn The callback function
+-- @realm client
+function vguihandler.RegisterCallback(name, fn)
```

Comment:
This allows only one function to be registered and it is also not stated in the doc